packages/react-instantsearch-hooks/src/lib/useInstantSearchApi.ts

```diff
@@ -31,11 +32,17 @@ export function useInstantSearchApi<TUiState extends UiState, TRouteState>(
 ) {
   const forceUpdate = useForceUpdate();
   const serverContext = useInstantSearchServerContext<TUiState, TRouteState>();
-  const serverState = useInstantSearchSSRContext();
+  const serverState = useInstantSearchSSRContext<TUiState, TRouteState>();
   const initialResults = serverState?.initialResults;
-  const searchRef = useRef<InstantSearch<TUiState, TRouteState> | null>(null);
   const prevPropsRef = useRef(props);
 
+  let searchRef = useRef<InstantSearch<TUiState, TRouteState> | null>(null);
+  // To prevent starting two instances with React Strict Mode + SSR, we use
+  // a ref from a parent.
+  if (serverState) {
+    searchRef = serverState.ssrSearchRef;
+  }
```

Comment:
Okay so basically with `InstantSearchSSRProvider`, in Strict Mode it creates two instances of InstantSearch and starts both without disposing the first.
We do have to start at first render else there's a hydration mismatch.

I spent quite a lot of time looking for a solution within this file but it doesn't seem possible so I added a `useRef` in the provider which we use instead.
The only drawback I can think of is it's no longer possible to have two instances of `InstantSearch` within the same `InstantSearchSSRProvider` context which I guess wouldn't make sense anyway